File: app/src/main/java/com/github/shadowsocks/TaskerActivity.java
package com.github.shadowsocks;


import android.content.res.Resources;
import android.content.res.TypedArray;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckedTextView;
import android.widget.Switch;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.github.shadowsocks.database.Profile;
import com.github.shadowsocks.utils.TaskerSettings;

import java.util.List;

import static com.github.shadowsocks.ShadowsocksApplication.app;

public class TaskerActivity extends AppCompatActivity {

    private TaskerSettings taskerOption;
    private Switch mSwitch;
    private ProfilesAdapter profilesAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.layout_tasker);

        profilesAdapter = new ProfilesAdapter();

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.app_name);
        toolbar.setNavigationIcon(R.drawable.ic_navigation_close);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        taskerOption = TaskerSettings.Companion.fromIntent(getIntent());
        mSwitch = findViewById(R.id.serviceSwitch);
        mSwitch.setChecked(taskerOption.getSwitchOn());
        RecyclerView profilesList = findViewById(R.id.profilesList);
        LinearLayoutManager lm = new LinearLayoutManager(this);
        profilesList.setLayoutManager(lm);
        profilesList.setItemAnimator(new DefaultItemAnimator());
        profilesList.setAdapter(profilesAdapter);

        if (taskerOption.getProfileId() >= 0) {
            int position = 0;
            List<Profile> profiles = profilesAdapter.profiles;
            for (int i = 0; i < profiles.size(); i++) {
                Profile profile = profiles.get(i);
                if (profile.getId() == taskerOption.getProfileId()) {
                    position = i + 1;
                    break;
                }
            }
            lm.scrollToPosition(position);
        }
    }

    private class ProfileViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private Profile item;
        private CheckedTextView text;

        public ProfileViewHolder(View view) {
            super(view);
            TypedArray typedArray = obtainStyledAttributes(new int[]{android.R.attr.selectableItemBackground});
            view.setBackgroundResource(typedArray.getResourceId(0, 0));
            typedArray.recycle();

            text = itemView.findViewById(android.R.id.text1);
            itemView.setOnClickListener(this);
        }

        public void bindDefault() {
            item = null;
            text.setText(R.string.profile_default);
            text.setChecked(taskerOption.getProfileId() < 0);
        }

        public void bind(Profile item) {
            this.item = item;
            text.setText(item.getName());
            text.setChecked(taskerOption.getProfileId() == item.getId());
        }

        @Override
        public void onClick(View v) {
            taskerOption.setSwitchOn(mSwitch.isChecked());
            taskerOption.setProfileId(item == null ? -1 : item.getId());
            setResult(RESULT_OK, taskerOption.toIntent(TaskerActivity.this));
            finish();
        }
    }

    private class ProfilesAdapter extends RecyclerView.Adapter<ProfileViewHolder> {

        private List<Profile> profiles;
        private String name;

        public ProfilesAdapter() {
            profiles = app.profileManager.getAllProfiles();

            String version = Build.VERSION.SDK_INT >= 21 ? "material" : "holo";
            name = "select_dialog_singlechoice_" + version;
        }

        public List<Profile> profiles() {
            return app.profileManager.getAllProfiles();
        }

        @Override
        public int getItemCount() {
            return 1 + profiles().size();
        }

        @Override
        public void onBindViewHolder(ProfileViewHolder vh, int i) {
            if (i == 0) {
                vh.bindDefault();
            } else {
                vh.bind(profiles().get(i - 1));
            }
        }

        @Override
        public ProfileViewHolder onCreateViewHolder(ViewGroup vg, int i) {
            View view = LayoutInflater.from(vg.getContext())
                    .inflate(Resources.getSystem().getIdentifier(name, "layout", "android"), vg, false);
            return new ProfileViewHolder(view);
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/TaskerActivity.kt
package com.bige0.shadowsocksr

import android.app.*
import android.content.res.*
import android.os.*
import android.view.*
import android.widget.*
import androidx.appcompat.app.*
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.*
import com.bige0.shadowsocksr.ShadowsocksApplication.Companion.app
import com.bige0.shadowsocksr.database.*
import com.bige0.shadowsocksr.utils.*

class TaskerActivity : AppCompatActivity()
{
	private lateinit var taskerOption: TaskerSettings
	private lateinit var mSwitch: Switch
	private lateinit var profilesAdapter: ProfilesAdapter

	override fun onCreate(savedInstanceState: Bundle?)
	{
		super.onCreate(savedInstanceState)
		setContentView(R.layout.layout_tasker)

		profilesAdapter = ProfilesAdapter()

		val toolbar = findViewById<Toolbar>(R.id.toolbar)
		toolbar.setTitle(R.string.app_name)
		toolbar.setNavigationIcon(R.drawable.ic_navigation_close)
		toolbar.setNavigationOnClickListener { finish() }

		taskerOption = TaskerSettings.fromIntent(intent)
		mSwitch = findViewById(R.id.serviceSwitch)
		mSwitch.isChecked = taskerOption.switchOn
		val profilesList = findViewById<RecyclerView>(R.id.profilesList)
		val lm = LinearLayoutManager(this)
		profilesList.layoutManager = lm
		profilesList.itemAnimator = DefaultItemAnimator()
		profilesList.adapter = profilesAdapter

		if (taskerOption.profileId >= 0)
		{
			var position = 0
			val profiles = profilesAdapter.getCurrentProfiles()
			for ((i, profile) in profiles.withIndex())
			{
				if (profile.id == taskerOption.profileId)
				{
					position = i + 1
					break
				}
			}
			lm.scrollToPosition(position)
		}
	}

	private inner class ProfileViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener
	{
		private var item: Profile? = null
		private val text: CheckedTextView

		init
		{
			val typedArray = obtainStyledAttributes(intArrayOf(android.R.attr.selectableItemBackground))
			view.setBackgroundResource(typedArray.getResourceId(0, 0))
			typedArray.recycle()

			text = itemView.findViewById(android.R.id.text1)
			itemView.setOnClickListener(this)
		}

		fun bindDefault()
		{
			item = null
			text.setText(R.string.profile_default)
			text.isChecked = taskerOption.profileId < 0
		}

		fun bind(item: Profile)
		{
			this.item = item
			text.text = item.name
			text.isChecked = taskerOption.profileId == item.id
		}

		override fun onClick(v: View)
		{
			taskerOption.switchOn = mSwitch.isChecked
			taskerOption.profileId = if (item == null) -1 else item!!.id
			setResult(Activity.RESULT_OK, taskerOption.toIntent(this@TaskerActivity))
			finish()
		}
	}

	private inner class ProfilesAdapter : RecyclerView.Adapter<ProfileViewHolder>()
	{
		private val name: String

		init
		{
			val version = if (Build.VERSION.SDK_INT >= 21) "material" else "holo"
			name = "select_dialog_singlechoice_$version"
		}

		fun getCurrentProfiles(): List<Profile>
		{
			return app.profileManager.allProfiles
		}

		override fun getItemCount(): Int
		{
			return 1 + getCurrentProfiles().size
		}

		override fun onBindViewHolder(vh: ProfileViewHolder, i: Int)
		{
			if (i == 0)
			{
				vh.bindDefault()
			}
			else
			{
				vh.bind(getCurrentProfiles()[i - 1])
			}
		}

		override fun onCreateViewHolder(vg: ViewGroup, i: Int): ProfileViewHolder
		{
			val view = LayoutInflater.from(vg.context)
				.inflate(Resources.getSystem().getIdentifier(name, "layout", "android"), vg, false)
			return ProfileViewHolder(view)
		}
	}
}
